lighthouse-core/lib/tracehouse/main-thread-tasks.js

```diff
@@ -426,38 +444,100 @@ class MainThreadTasks {
       case 'v8.compile':
       case 'EvaluateScript':
       case 'FunctionCall':
-        taskURLs = [argsData.url].concat(stackFrameURLs);
+        taskURLs = [argsData.url, frameURL];
         break;
       case 'v8.compileModule':
-        taskURLs = [task.event.args.fileName].concat(stackFrameURLs);
+        taskURLs = [task.event.args.fileName];
         break;
       case 'TimerFire': {
         /** @type {string} */
         // @ts-ignore - timerId exists when name is TimerFire
         const timerId = task.event.args.data.timerId;
         const timerInstallerTaskNode = priorTaskData.timers.get(timerId);
         if (!timerInstallerTaskNode) break;
-        taskURLs = timerInstallerTaskNode.attributableURLs.concat(stackFrameURLs);
+        taskURLs = timerInstallerTaskNode.attributableURLs;
         break;
       }
+      case 'ParseHTML':
+        taskURLs = [argsData.url, frameURL];
+        break;
+      case 'ParseAuthorStyleSheet':
+        taskURLs = [argsData.styleSheetUrl, frameURL];
+        break;
+      case 'UpdateLayoutTree':
+      case 'Layout':
+      case 'Paint':
+        // If we had a specific frame we were updating, just attribute it to that frame.
+        if (frameURL) {
+          taskURLs = [frameURL];
+          break;
+        }
+
+        // Otherwise, sometimes Chrome will split layout into separate toplevel task after things have settled.
+        // In this case we want to attribute the work to the prior task.
+        // Inherit from previous task only if we don't have any task data set already.
+        if (allURLsInTree.length) break;
+        taskURLs = priorTaskData.lastTaskURLs;
+        break;
+      case 'XHRReadyStateChange':
+      case 'XHRLoad':
+        // Inherit from task that issued the XHR
+        const xhrUrl = argsData.url;
+        const readyState = argsData.readyState;
+        if (!xhrUrl || (typeof readyState === 'number' && readyState !== 4)) break;
+        const xhrRequesterTaskNode = priorTaskData.xhrs.get(xhrUrl);
+        if (!xhrRequesterTaskNode) break;
+        taskURLs = xhrRequesterTaskNode.attributableURLs;
+        break;
       default:
-        taskURLs = stackFrameURLs;
+        taskURLs = [];
         break;
     }
 
     /** @type {string[]} */
     const attributableURLs = Array.from(parentURLs);
-    for (const url of taskURLs) {
+    for (const url of taskURLs.concat(stackFrameURLs)) {
```

Comment:
is `[...taskURLs, ...stackFrameURLs]` slower?

Reply:
if i remember right, spreading is actually quite a bit faster.
(js microbenchmarks! my favorite. 😢 )

Reply:
tis true, ~33% faster :)

https://jsperf.com/spread-vs-concat-vs-push